### DISCUSSION TITLE
Equavalent of `pip install -e .`?

### DISCUSSION THREAD

**Discussion by asmith26** [prefix-dev/pixi#1463] [Q&A] [answered]
I've ran through the [basic usage tutorial](https://pixi.sh/latest/basic_usage/), but noticed that the `pixi-hello-world` project/package does not appear.

```bash
$ pixi list       
Package           Version   Build               Size       Kind   Source
_libgcc_mutex     0.1       conda_forge         2.5 KiB    conda  _libgcc_mutex-0.1-conda_forge.tar.bz2
_openmp_mutex     4.5       2_gnu               23.1 KiB   conda  _openmp_mutex-4.5-2_gnu.tar.bz2
bzip2             1.0.8     hd590300_5          248.3 KiB  conda  bzip2-1.0.8-hd590300_5.conda
ca-certificates   2024.2.2  hbcca054_0          151.8 KiB  conda  ca-certificates-2024.2.2-hbcca054_0.conda
ld_impl_linux-64  2.40      hf3520f5_1          691.3 KiB  conda  ld_impl_linux-64-2.40-hf3520f5_1.conda
libexpat          2.6.2     h59595ed_0          72 KiB     conda  libexpat-2.6.2-h59595ed_0.conda
libffi            3.4.2     h7f98852_5          56.9 KiB   conda  libffi-3.4.2-h7f98852_5.tar.bz2
libgcc-ng         13.2.0    h77fa898_7          757.6 KiB  conda  libgcc-ng-13.2.0-h77fa898_7.conda
libgomp           13.2.0    h77fa898_7          412.4 KiB  conda  libgomp-13.2.0-h77fa898_7.conda
libnsl            2.0.1     hd590300_0          32.6 KiB   conda  libnsl-2.0.1-hd590300_0.conda
libsqlite         3.45.3    h2797004_0          839.7 KiB  conda  libsqlite-3.45.3-h2797004_0.conda
libuuid           2.38.1    h0b41bf4_0          32.8 KiB   conda  libuuid-2.38.1-h0b41bf4_0.conda
libxcrypt         4.4.36    hd590300_1          98 KiB     conda  libxcrypt-4.4.36-hd590300_1.conda
libzlib           1.2.13    h4ab18f5_6          60.1 KiB   conda  libzlib-1.2.13-h4ab18f5_6.conda
ncurses           6.5       h59595ed_0          866.7 KiB  conda  ncurses-6.5-h59595ed_0.conda
openssl           3.3.0     h4ab18f5_3          2.8 MiB    conda  openssl-3.3.0-h4ab18f5_3.conda
python            3.12.3    hab00c5b_0_cpython  30.5 MiB   conda  python-3.12.3-hab00c5b_0_cpython.conda
readline          8.2       h8228510_1          274.9 KiB  conda  readline-8.2-h8228510_1.conda
tk                8.6.13    noxft_h4845f30_101  3.2 MiB    conda  tk-8.6.13-noxft_h4845f30_101.conda
tzdata            2024a     h0c530f3_0          117 KiB    conda  tzdata-2024a-h0c530f3_0.conda
xz                5.2.6     h166bdaf_0          408.6 KiB  conda  xz-5.2.6-h166bdaf_0.tar.bz2
```

Is there a way to install the pixi package being developed into the current pixi environment? (I'm using a pixi.toml, so unfortunately `pip install -e .` does not seem to work)

Thanks!

**Comment by ruben-arts** [comment]
If you want to do editable installs you need to add them as `pypi-dependencies`. This requires a build system so if you follow the [python tutorial](https://pixi.sh/latest/tutorials/python/#whats-in-the-environment) you should be good to go!

e.g.
```toml
[pypi-dependencies]
pixi-hello-world = { path = ".", editable = true}
```

**Reply by DeadlySquad13** [reply]
Hello, thanks for your suggestion! I`ve encountered the same issue - the code snippet you provided helped with configuring *local* pixi environment. But is it possible to install package as editable **globally**?

I'm developing a cli tool and currently I'm at stage when all functionality is written and it's only left to battle-test the package. In situations like this I would run `pip install -e .` (in my case it's actually `pipx install -e .` because of Arch policy) and use the tool in everyday life / fix small bugs simultaneously. Unfortunately, I found that I have to copy-paste all the `pixi.toml` configuration into `pyproject.toml` or `setup.cfg` to make it work - otherwise globally installed tool wouldn't find packages declared in pixi.

I tried my best reading docs and searching through issues but to no avail. Maybe there's an alternative means to achieve this workflow?

Repo: https://github.com/DeadlySquad13/InfoField__FileSystemTaxonomy/tree/refactor/pixi-rewrite
Sorry for the mess: it's quite a lot of duplicate code across `pixi.toml`, `setup.cfg` and `pyproject.toml` because I was trying to solve the issue.

**Reply by altaic** [reply]
It’s been nearly a year with many developments. Is it possible to install a project globally? Or is there a workaround that does not require external servers?

**Reply by DeadlySquad13** [reply]
> It’s been nearly a year with many developments. Is it possible to install a project globally? Or is there a workaround that does not require external servers?

It was possible to install it globally when I was writing the comment. There's even a section for that in docs. The problem was specifically in **editable** global install.

**Reply by altaic** [reply]
> It was possible to install it globally when I was writing the comment. There's even a section for that in docs. The problem was specifically in **editable** global install.

Sorry, that's what I was referring to. Is globally installing an editable project currently possible, or is there a workaround that doesn't involve publishing to a conda channel on a server?

For instance, a workaround I came up with for python packages is to globally install it, then edit libs in that environment's `site-packages` dir. Not very flexible, but it works in a pinch.

**Comment by asmith26** [comment, marked as answer]
To summarize what works for me:

1. Create a pyproject.toml project: `pixi init --pyproject pixi-hello-world`
2. Create a package folder structure using the common convention e.g. `src/pixi_hello_world/main.py`
3. Update pixi environment `pixi install`
4. Check `pixi_hello_world` package added with `pixi list`

**Comment by jonmoore** [comment]
Given this, it seems like most people using pixi for Python will be better off using a pyproject.toml rather than pixi.toml, and so it would be helpful to adjust https://pixi.sh/latest/basic_usage/ to recommend that.

**Reply by DeadlySquad13** [reply]
It seems this way because I just ended up replicating `pixi.toml` in `pyproject.toml`.
